### PR TITLE
AB#40963 Store draft files in database (rather than JSON fles)

### PR DESCRIPTION
**Pull request by jamesgriff**

### CODE REVIEW

File: GenderPayGap.WebUI/Startup.cs [hunk 1]
```diff
@@ -217,7 +217,7 @@ public IContainer BuildContainerIoC(IServiceCollection services)
             builder.RegisterType<UpdateFromCompaniesHouseService>().As<UpdateFromCompaniesHouseService>().InstancePerLifetimeScope();
 
             // register web ui services
-            builder.RegisterType<DraftFileBusinessLogic>().As<IDraftFileBusinessLogic>().SingleInstance();
+            builder.RegisterType<DraftFileBusinessLogic>().As<IDraftFileBusinessLogic>().InstancePerLifetimeScope();
```
Comment (aidayounesi) [inline GenderPayGap.WebUI/Startup.cs:221]: Neat

File: GenderPayGap.BusinessLogic/Services/DraftFileBusinessLogic.cs [hunk 2]
```diff
@@ -294,14 +295,14 @@ private async Task WriteInDbAndTimestampAsync(Draft resultingDraft, long userIdR
...
             await SetDraftReturnFromDraft(draftFromDb, resultingDraft);
         }
 
-        private async Task<DraftReturn> GetDraftReturnFromDatabase(long organisationId, int snapshotYear, DraftReturnStatus draftReturnStatus)
+        private DraftReturn GetDraftReturnFromDatabase(long organisationId, int snapshotYear, DraftReturnStatus draftReturnStatus)
         {
-            return await dataRepository.GetAll<DraftReturn>()
-                .FirstOrDefaultAsync(
+            return dataRepository.GetAll<DraftReturn>()
+                .FirstOrDefault(
```
Comment (aidayounesi) [inline GenderPayGap.BusinessLogic/Services/DraftFileBusinessLogic.cs]: Awesome :)

File: GenderPayGap.BusinessLogic/Services/DraftFileBusinessLogic.cs [hunk 3]
```diff
@@ -302,117 +309,152 @@ private Draft CastDatabaseDraftReturnToDraft(DraftReturn draftReturn)
...
+            List<DraftReturn> matchingReturns = dataRepository.GetAll<DraftReturn>()
+                .Where(dr => dr.OrganisationId == draftToSave.OrganisationId)
+                .Where(dr => dr.SnapshotYear == draftToSave.SnapshotYear)
+                .Where(dr => dr.DraftReturnStatus == draftToSave.DraftReturnStatus)
+                .ToList();
+
+            foreach (DraftReturn matchingReturn in matchingReturns)
+            {
+                dataRepository.Delete(matchingReturn);
             }
```
Comment (aidayounesi) [inline GenderPayGap.BusinessLogic/Services/DraftFileBusinessLogic.cs:370]: I like your approach here

File: GenderPayGap.BusinessLogic/Services/DraftFileBusinessLogic.cs [hunk 4]
```diff
@@ -153,17 +151,17 @@ public async Task CommitDraftAsync(Draft draftToDiscard)
 
         #region private methods
 
-        private async Task<Draft> GetDraftOrCreateAsync(Draft resultingDraft, long userIdRequestingAccess)
+        private Draft GetDraftOrCreateAsync(Draft resultingDraft, long userIdRequestingAccess)
```
Comment (aidayounesi) [inline GenderPayGap.BusinessLogic/Services/DraftFileBusinessLogic.cs:118]: It worth renaming any method with "Async" in its name

File: GenderPayGap.UnitTests/GenderPayGap.WebUI.Tests/Services/ViewingServiceTests.cs [hunk 5]
```diff
@@ -120,6 +120,7 @@ public async Task GetSectorOptions_Marks_given_sector_chars_as_checked_or_not_ch
         }
 
         [Test]
+        [Ignore("This breaks in late-June and/or 1st-January each year because the date calculation in the test is wrong")]
```
Comment (aidayounesi) [inline GenderPayGap.UnitTests/GenderPayGap.WebUI.Tests/Services/ViewingServiceTests.cs:123]: 👍

File: GenderPayGap.Database/Migrations/20200625092812_AddDraftReturnTable.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    public partial class AddDraftReturnTable : Migration
+    {
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable(
+                name: "DraftReturns",
+                columns: table => new
+                {
+                    DraftReturnId = table.Column<long>(nullable: false)
+                        .Annotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn),
```
Comment (aidayounesi) [inline GenderPayGap.Database/Migrations/20200625092812_AddDraftReturnTable.cs:16]: Is it going to work after migration?

File: GenderPayGap.WebUI/Controllers/Admin/AdminMigrateDraftReturnsToDatabaseController.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
+﻿using System;
```
Comment (aidayounesi) [inline GenderPayGap.WebUI/Controllers/Admin/AdminMigrateDraftReturnsToDatabaseController.cs:1]: 🥇 💯

**Comment by aidayounesi** [comment]
Thank you James, it looks great

File: GenderPayGap.BusinessLogic/Services/DraftFileBusinessLogic.cs (hunk 4 above)
Comment (jamesgriff) [reply GenderPayGap.BusinessLogic/Services/DraftFileBusinessLogic.cs:118]: Great spot! I'll change these

File: GenderPayGap.Database/Migrations/20200625092812_AddDraftReturnTable.cs (hunk 6 above)
Comment (jamesgriff) [reply GenderPayGap.Database/Migrations/20200625092812_AddDraftReturnTable.cs:16]: Probably not - I think we'll have to do another round of "squash all the database migrations"

**Pull request merged**